parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/RecordProcessor.java

```diff
@@ -0,0 +1,48 @@
+package io.confluent.parallelconsumer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.util.List;
+
+/**
+ * Types of user functions used for processing records.
+ */
+public interface RecordProcessor {
+
+    /**
+     * Process a Kafka {@link ConsumerRecord} via {@link PollContext} instances.
+     */
+    @FunctionalInterface
+    interface PollConsumer<K, V> extends java.util.function.Consumer<PollContext<K, V>> {
+
+        /**
+         * Process a Kafka {@link ConsumerRecord} via {@link PollContext} instances.
+         * <p>
+         * User can throw a {@link PCRetriableException}, if an issue is and PC should handle the process of retrying it
+         * later. If an exception is thrown that doesn't extend {@link PCRetriableException}, the error will be logged
+         * at {@code WARN} level.
+         *
+         * @param records the Kafka records to process
+         * @see PCRetriableException
+         * @see ParallelConsumerOptions#getRetryDelayProvider()
+         * @see ParallelConsumerOptions#getDefaultMessageRetryDelay()
+         */
+        void accept(PollContext<K, V> records);
+    }
+
+    @FunctionalInterface
+    interface PollConsumerAndProducer<K, V> extends java.util.function.Function<PollContext<K, V>, List<ProducerRecord<K, V>>> {
```

Comment:
Staying with PollConsumerAndProducer ? Would MessageTransformer or Processor or Mapper sound better given the messages in / out nature?

Reply:
yeah - other thing is, if it affects your ideas - is it can produce many messages, not just one, from a single input

Reply:
Would it be worth going to KStream DSL naming? have a map and flatmap where map takes 1 in - 1 out, and flatmap allows 1->many ? The typing needs fixed to allow transform of key / value anyway - so maybe can be done in separate PR that deals with it. Or Transform (with or without flatTransform).
Thinking about it - "Transformer" sounds best to me. 
Are side effects allowed / expected in this scenario (i.e. external system calls for processing / enrichment)? If so something neutral like Processor would be ok i think.